order-book/src/actors/executor.rs

```diff
@@ -0,0 +1,145 @@
+    FixedInput,
+    FixedOutput,
+}
+
+#[derive(TypeAbi, TopEncode, TopDecode, NestedEncode, NestedDecode, ManagedVecItem)]
+pub struct SwapOperationType<M: ManagedTypeApi> {
+    pub pair_address: ManagedAddress<M>,
+    pub endpoint_name: RouterEndpointName,
+    pub output_token_id: TokenIdentifier<M>,
+}
+
+#[multiversx_sc::module]
+pub trait ExecutorModule:
+    crate::external_sc_interactions::router::RouterActionsModule
+    + crate::storage::order::OrderModule
+    + crate::storage::common_storage::CommonStorageModule
+    + crate::events::EventsModule
+    + crate::pause::PauseModule
+{
+    /// args are pairs of (order ID, token amount to swap, vec of SwapOperationType - which will be passed to router when swapping)
+    #[endpoint(executeOrders)]
+    fn execute_orders(
+        &self,
+        args: MultiValueEncoded<
+            MultiValue3<OrderId, BigUint, ManagedVec<SwapOperationType<Self::Api>>>,
+        >,
+    ) -> MultiValueEncoded<SwapStatus> {
+        let caller = self.blockchain().get_caller();
+        require!(
+            self.executor_whitelist().contains(&caller),
+            "Not in executor whitelist"
+        );
+
+        let mut swap_statuses = MultiValueEncoded::new();
+        for arg in args {
+            let (order_id, input_token_amount, swap_path) = arg.into_tuple();
+            let is_valid = self.validate_input(order_id, &input_token_amount, &swap_path);
+            if !is_valid {
+                swap_statuses.push(SwapStatus::InvalidInput);
+
+                continue;
+            }
+
+            let mut order = self.orders(order_id).get();
+            let opt_tokens_out = self.execute_swap(&order, &input_token_amount, &swap_path);
+            match opt_tokens_out {
+                OptionalValue::Some(payment) => {
+                    self.update_order_after_success(
+                        order_id,
+                        &mut order,
+                        &payment,
+                        &input_token_amount,
+                    );
+                    self.distribute_tokens();
+
+                    swap_statuses.push(SwapStatus::Success);
+                }
+                OptionalValue::None => {
+                    swap_statuses.push(SwapStatus::Fail);
+                }
+            }
+        }
+
+        swap_statuses
+    }
+
+    /// returns `true` if input is valid, `false` otherwise
+    #[must_use]
+    fn validate_input(
+        &self,
+        order_id: OrderId,
+        token_amount: &BigUint,
+        swap_path: &ManagedVec<SwapOperationType<Self::Api>>,
+    ) -> bool {
+        if token_amount == &0 {
+            return false;
+        }
+
+        let order_mapper = self.orders(order_id);
+        if order_mapper.is_empty() {
+            return false;
+        }
+
+        let order = order_mapper.get();
+        let current_time = self.blockchain().get_block_timestamp();
+        if order.expiration_timestamp >= current_time {
```

Comment:
Wrong condition, should be the other way around.

Reply:
Fixed.